src/main/java/org/opensearch/neuralsearch/processor/TextEmbeddingProcessor.java

```diff
@@ -0,0 +1,269 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Before querying with neural search, all the documents should be ingested in the form of embedded vectors. Leaving the embedding process
+ * to the user offline will raise the bar of use, thus we create this processor for user input document text embedding.
+ */
+@Log4j2
+public class TextEmbeddingProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "text_embedding";
+    public static final String MODEL_ID_FIELD = "model_id";
+    public static final String FIELD_MAP_FIELD = "field_map";
+
+    private static final String LIST_TYPE_NESTED_MAP_KEY = "knn";
+
+    private final String modelId;
+
+    private final Map<String, Object> fieldMap;
+
+    private final MLCommonsClientAccessor mlCommonsClientAccessor;
+
+    public TextEmbeddingProcessor(
+        String tag,
+        String description,
+        String modelId,
+        Map<String, Object> fieldMap,
+        MLCommonsClientAccessor clientAccessor
+    ) {
+        super(tag, description);
+        if (StringUtils.isBlank(modelId)) throw new IllegalArgumentException("model_id is null, can not process it");
+        if (fieldMap == null || fieldMap.size() == 0 || checkEmbeddingConfigNotValid(fieldMap)) throw new IllegalArgumentException(
+            "Unable to create the TextEmbedding processor as field_map is null or empty."
+        );
+
+        this.modelId = modelId;
+        this.fieldMap = fieldMap;
+        this.mlCommonsClientAccessor = clientAccessor;
+    }
+
+    private boolean checkEmbeddingConfigNotValid(Map<String, Object> fieldMap) {
+        return fieldMap.entrySet()
+            .stream()
+            .anyMatch(x -> StringUtils.isBlank(x.getKey()) || Objects.isNull(x.getValue()) || StringUtils.isBlank(x.getValue().toString()));
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) {
+        validateEmbeddingFieldsType(ingestDocument, fieldMap);
+        Map<String, Object> knnMap = buildKnnMap(ingestDocument, fieldMap);
+        try {
+            List<List<Float>> vectors = mlCommonsClientAccessor.blockingInferenceSentences(this.modelId, createInferenceList(knnMap));
+            processResponse(ingestDocument, knnMap, vectors);
+        } catch (ExecutionException | InterruptedException e) {
+            log.error("Text embedding processor failed with exception: " + e.getMessage(), e);
+            throw new RuntimeException("Text embedding processor failed with exception", e);
+        }
+        log.info("Text embedding completed, returning ingestDocument!");
+        return ingestDocument;
+    }
+
+    void processResponse(IngestDocument ingestDocument, Map<String, Object> knnMap, List<List<Float>> vectors) {
+        Objects.requireNonNull(vectors, "embedding failed, inference returns null result!");
+        log.info("Text embedding result fetched, starting build vector output!");
```

Comment:
qq: If this method is executed on each doc ingest, I dont think we should log.info -> itd be better to switch to debug. What happens if 1M documents are ingested?

Reply:
Changed log to debug level.